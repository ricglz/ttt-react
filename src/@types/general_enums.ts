export const enum Cell {
  X = 1,
  O = -1,
  NONE = 0,
}
export const enum Player {
  PLAYER_1 = 'X',
  PLAYER_2 = 'O',
}

export const enum Difficulty {
  EASY = 1,
  NORMAL = 2,
  HARD = 3,
}
