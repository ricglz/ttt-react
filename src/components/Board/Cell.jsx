import React from 'react';
import PropTypes from 'prop-types';

const currentValue = (boardGame, boardNum, cellNum) => boardGame[boardNum][cellNum];

function drawValue(boardGame, boardNum, cellNum) {
  switch (currentValue(boardGame, boardNum, cellNum)) {
    case 1:
      return 'X';
    case -1:
      return 'O';
    default:
      return '';
  }
}

const emptyFunction = () => {
  // do nothing
};

const Cell = ({
  boardGame, boardNum, cellNum, handleClick,
}) => {
  const value = drawValue(boardGame, boardNum, cellNum);
  const onClick = React.useCallback(() => {
    handleClick(boardNum, cellNum);
  }, [boardNum, cellNum, handleClick]);
  return (
    <div
      className="col-4 box"
      onClick={onClick}
      onKeyDown={emptyFunction}
      role="button"
      tabIndex="-1"
    >
      <p className={value.toLowerCase()}>{value}</p>
    </div>
  );
};

Cell.propTypes = {
  boardGame: PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.number)).isRequired,
  boardNum: PropTypes.number.isRequired,
  cellNum: PropTypes.number.isRequired,
  handleClick: PropTypes.func.isRequired,
};

export default Cell;
