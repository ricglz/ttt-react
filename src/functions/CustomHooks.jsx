import React from 'react';
import {
  isOccupied,
  theresAWinner,
  initialState,
  alertWinner,
} from './HelperFunctions';
import makeMove, {
  cleanVariables,
  aiMadeAMove,
} from './Ai';

const CONSTANTS = {
  PLAYER1: 'X',
  PLAYER2: 'O',
};

export function useGameHooks(game, setGame) {
  const {
    boardGame, currentBoard, currentPlayer,
  } = game;

  const canClick = React.useCallback((board, id) => {
    const currentValue = boardGame[board][id];
    if (isOccupied(currentValue)) return false;
    return board === currentBoard || currentBoard === -1;
  }, [boardGame, currentBoard]);

  const pvpMove = React.useCallback((boardCopy, newMoveNumber, id) => {
    setGame({
      ...game,
      boardGame: boardCopy,
      moveNumber: newMoveNumber,
      currentBoard: id,
      currentPlayer:
        currentPlayer === CONSTANTS.PLAYER1
          ? CONSTANTS.PLAYER2
          : CONSTANTS.PLAYER1,
    });
  }, [setGame, game, currentPlayer]);

  const newGame = React.useCallback(() => {
    cleanVariables();
    setGame({ ...game, ...initialState() });
  }, [setGame, game]);

  return [canClick, pvpMove, newGame];
}

export function useAIHooks({
  moveNumber,
  selectedOption,
  setGame,
  changeScore,
  newGame,
  setSelectedOption,
  game,
}) {
  const aiMove = React.useCallback((boardCopy, id, newMoveNumber) => {
    const board = boardCopy;
    const difficulty = selectedOption == null ? 1 : selectedOption.value;
    const move = makeMove(board[id], id, difficulty);
    board[id][move] = -1;
    const winner = theresAWinner(board[id]);
    if (winner) {
      alertWinner(winner);
      changeScore(winner);
      newGame();
    } else {
      aiMadeAMove(id, board[id]);
      setGame({
        ...game,
        boardGame: board,
        moveNumber: newMoveNumber + 1,
        currentBoard: move,
      });
    }
  }, [changeScore, newGame, selectedOption, setGame, game]);

  const handleChange = React.useCallback((newSelectedOption) => {
    setSelectedOption(newSelectedOption);
    if (moveNumber >= 0) {
      newGame();
    }
  }, [setSelectedOption, moveNumber, newGame]);

  return [aiMove, handleChange];
}
