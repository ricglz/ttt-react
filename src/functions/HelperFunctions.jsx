import { NotificationManager } from 'react-notifications';

let boardCopy = null;

export function alertWinner(winner) {
  const winnerLabel = winner === 1 ? 'X' : 'O';
  NotificationManager.info(`Player ${winnerLabel} has won.`);
}

export function emptyArray() {
  return [0, 0, 0, 0, 0, 0, 0, 0, 0];
}

export function isOccupied(value) {
  return value !== 0;
}

function allThree(first, second, third) {
  const firstValue = boardCopy[first];
  const secondValue = boardCopy[second];
  const thirdValue = boardCopy[third];
  if (
    firstValue === secondValue
    && secondValue === thirdValue
    && isOccupied(firstValue)
  ) {
    return firstValue;
  }
  return null;
}

function columnWin() {
  return allThree(0, 3, 6) || allThree(1, 4, 7) || allThree(2, 5, 8);
}

function rowWin() {
  return allThree(0, 1, 2) || allThree(3, 4, 5) || allThree(6, 7, 8);
}

function diagonalWin() {
  return allThree(0, 4, 8) || allThree(2, 4, 6);
}

export function theresAWinner(board) {
  boardCopy = board;
  return columnWin() || rowWin() || diagonalWin();
}

export function newBoard() {
  return [
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
  ];
}

export function initialState() {
  return {
    boardGame: newBoard(),
    currentPlayer: 'X',
    moveNumber: 0,
    currentBoard: -1,
  };
}

export function fbInitialState(hostUid, hostName) {
  return {
    boardGame: newBoard(),
    currentPlayer: 'X',
    moveNumber: 0,
    currentBoard: -1,
    oWins: 0,
    xWins: 0,
    hostUid,
    hostName,
    guestUid: -1,
    nextPlayerUid: hostUid,
    timestamp: Date.now(),
  };
}
