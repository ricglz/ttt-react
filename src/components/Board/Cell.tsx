import { useCallback } from "react";

import type { BigBoard, GeneralBoardIndex } from "types/general";
import { Player, Cell as CellEnum } from "types/general_enums";
import type { BoardProps } from "./Board";

type Args = {
  boardGame: BigBoard;
  boardNum: GeneralBoardIndex;
  cellNum: GeneralBoardIndex;
};

const currentValue = ({ boardGame, boardNum, cellNum }: Args) =>
  boardGame[boardNum][cellNum];

function drawValue(args: Args) {
  switch (currentValue(args)) {
    case CellEnum.X:
      return Player.PLAYER_1;
    case CellEnum.O:
      return Player.PLAYER_2;
    default:
      return "";
  }
}

const emptyFunction = () => {};

interface Props extends BoardProps {
  cellNum: GeneralBoardIndex;
}

const Cell = ({ boardGame, boardNum, cellNum, handleClick }: Props) => {
  const value = drawValue({ boardGame, boardNum, cellNum });
  const onClick = useCallback(() => {
    handleClick(boardNum, cellNum);
  }, [boardNum, cellNum, handleClick]);
  return (
    <div
      className="col-4 box"
      onClick={onClick}
      onKeyDown={emptyFunction}
      role="button"
      tabIndex={-1}
    >
      <p className={value.toLowerCase()}>{value}</p>
    </div>
  );
};

export default Cell;
