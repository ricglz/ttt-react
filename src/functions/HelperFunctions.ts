import { NotificationManager } from 'react-notifications';
import type {
  BigBoard, Board, GeneralBoardIndex, NumberBoard,
} from '../@types/general';
import { Cell, Player } from '../@types/general_enums';

export const getNextPlayer = (player: Player) => (
  player === Player.PLAYER_1 ? Player.PLAYER_2 : Player.PLAYER_1
);

export const getPlayerCellValue = (player: Player) => (
  player === Player.PLAYER_1 ? Cell.X : Cell.O
);

export function alertWinner(winner: Cell) {
  const winnerLabel = winner === Cell.X ? Player.PLAYER_1 : Player.PLAYER_2;
  NotificationManager.info(`Player ${winnerLabel} has won.`);
}

export function alertError(err: Error) {
  NotificationManager.error(err.message);
}

export function emptyArray(): NumberBoard {
  return [0, 0, 0, 0, 0, 0, 0, 0, 0];
}

export function isOccupied(value: Cell) {
  return value !== Cell.NONE;
}

function allThree(
  first: GeneralBoardIndex,
  second: GeneralBoardIndex,
  third: GeneralBoardIndex,
  board: Board,
) {
  const firstValue = board[first];
  const secondValue = board[second];
  const thirdValue = board[third];
  if (
    firstValue === secondValue
    && secondValue === thirdValue
    && isOccupied(firstValue)
  ) {
    return firstValue;
  }
  return null;
}

function columnWin(board: Board) {
  return allThree(0, 3, 6, board) || allThree(1, 4, 7, board) || allThree(2, 5, 8, board);
}

function rowWin(board: Board) {
  return allThree(0, 1, 2, board) || allThree(3, 4, 5, board) || allThree(6, 7, 8, board);
}

function diagonalWin(board: Board) {
  return allThree(0, 4, 8, board) || allThree(2, 4, 6, board);
}

export function theresAWinner(board: Board) {
  return columnWin(board) || rowWin(board) || diagonalWin(board);
}

export function newBoard(): BigBoard {
  return [
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
  ];
}

export type CurrentBoard = GeneralBoardIndex | -1;

interface BaseGame {
  boardGame: BigBoard;
  currentPlayer: Player;
  moveNumber: number;
  currentBoard: CurrentBoard;
}

export function initialState(): BaseGame {
  return {
    boardGame: newBoard(),
    currentPlayer: Player.PLAYER_1,
    moveNumber: 0,
    currentBoard: -1,
  };
}

export interface FirebaseGame extends BaseGame {
  oWins: number;
  xWins: number;
  hostUid: string;
  hostName: string;
  guestUid: string;
  nextPlayerUid: string;
  timestamp: number;
}

export function fbInitialState(hostUid: string, hostName: string): FirebaseGame {
  return {
    ...initialState(),
    oWins: 0,
    xWins: 0,
    hostUid,
    hostName,
    guestUid: '-1',
    nextPlayerUid: hostUid,
    timestamp: Date.now(),
  };
}

export type Game = BaseGame | FirebaseGame;
