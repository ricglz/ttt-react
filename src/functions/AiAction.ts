import type { GeneralBoardIndex, Board, NumberBoard } from "types/general";
import { Cell, Difficulty } from "types/general_enums";

import { isOccupied } from "./HelperFunctions";

type AiArgs = {
  pos: GeneralBoardIndex;
  currentDifficulty: Difficulty;
  boardCopy: Board;
  currentBoard: GeneralBoardIndex;
  amountOccupied: NumberBoard;
  avoidBox: NumberBoard;
};
type ArgsWithoutDifficulty = Omit<AiArgs, "currentDifficulty">;

function someOneCouldWin(value1: Cell, value2: Cell, player = false) {
  const hopedValue = player ? Cell.X : Cell.O;
  return (
    (value1 === hopedValue && !isOccupied(value2)) ||
    (value2 === hopedValue && !isOccupied(value1))
  );
}

const aiCouldWin = (value1: Cell, value2: Cell) =>
  someOneCouldWin(value1, value2);
const playerCouldWin = (value1: Cell, value2: Cell) =>
  someOneCouldWin(value1, value2, true);

// Identifies which letter are the one near it and if it would be more valuable
// to do the move or not
function whatAreBoth(
  column1: GeneralBoardIndex,
  column2: GeneralBoardIndex,
  { boardCopy }: ArgsWithoutDifficulty
) {
  const value1 = boardCopy[column1];
  const value2 = boardCopy[column2];
  if (value1 === value2) {
    if (value1 === Cell.X) return 50;
    if (value2 === Cell.O) return 9000;
  }
  let value = 0;
  if (playerCouldWin(value1, value2)) value += 15;
  if (aiCouldWin(value1, value2)) value += 40;
  return value;
}

// Checks if the position chosen could stop a winning of the oponent or if
// itself could win in diagonal
function extraValueDiagonal(
  div: number,
  pos: GeneralBoardIndex,
  args: ArgsWithoutDifficulty
) {
  if (pos === 1 || pos === 3 || pos === 5 || pos === 7) return 0;
  if (pos === 4) return whatAreBoth(0, 8, args) + whatAreBoth(2, 6, args);
  if (div !== 0) {
    if (pos === 8) return whatAreBoth(0, 4, args);
    return whatAreBoth(2, 4, args);
  }
  if (pos === 0) return whatAreBoth(4, 8, args);
  return whatAreBoth(4, 6, args);
}

// Checks if the position chosen could stop a winning of the oponent or if itself could win in a row
function extraValueRow(div: number, mod: number, args: ArgsWithoutDifficulty) {
  switch (div) {
    case 0:
      if (mod === 0) return whatAreBoth(1, 2, args);
      if (mod === 1) return whatAreBoth(0, 2, args);
      return whatAreBoth(0, 1, args);
    case 1:
      if (mod === 0) return whatAreBoth(4, 5, args);
      if (mod === 1) {
        return whatAreBoth(3, 5, args);
      }
      return whatAreBoth(3, 4, args);
    default:
      if (mod === 0) return whatAreBoth(7, 8, args);
      if (mod === 1) return whatAreBoth(6, 8, args);
      return whatAreBoth(6, 7, args);
  }
}

// Checks if the position chosen could stop a winning of the oponent or if
// itself could win in a column
function extraValueColumn(
  div: number,
  mod: number,
  args: ArgsWithoutDifficulty
) {
  switch (mod) {
    case 0:
      if (div === 0) return whatAreBoth(3, 6, args);
      if (div === 1) return whatAreBoth(0, 6, args);
      return whatAreBoth(0, 3, args);
    case 1:
      if (div === 0) return whatAreBoth(4, 7, args);
      if (div === 1) return whatAreBoth(1, 7, args);
      return whatAreBoth(1, 4, args);
    default:
      if (div === 0) return whatAreBoth(5, 8, args);
      if (div === 1) return whatAreBoth(2, 8, args);
      return whatAreBoth(2, 5, args);
  }
}

function positiveValues(args: ArgsWithoutDifficulty) {
  const { pos } = args;
  const div = Math.floor(pos / 3);
  const mod = pos % 3;
  return (
    extraValueColumn(div, mod, args) +
    extraValueRow(div, mod, args) +
    extraValueDiagonal(div, pos, args)
  );
}

const recursiveMovement = ({ pos, currentBoard }: ArgsWithoutDifficulty) =>
  pos === currentBoard ? 20 : 0;

function negativeValues(args: ArgsWithoutDifficulty) {
  const { amountOccupied, avoidBox, pos } = args;
  return recursiveMovement(args) + amountOccupied[pos] + avoidBox[pos];
}

function getValue({ currentDifficulty, ...rest }: AiArgs) {
  let value = positiveValues(rest);
  if (currentDifficulty === Difficulty.HARD) value -= negativeValues(rest);
  return value;
}

class AiAction {
  pos: GeneralBoardIndex;

  value: number;

  constructor(args: AiArgs) {
    this.pos = args.pos;
    this.value = getValue(args);
  }
}

export default AiAction;
