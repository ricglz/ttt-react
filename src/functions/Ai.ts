import type { Board, GeneralBoardIndex, NumberBoard } from "types/general";

import AiAction from "./AiAction";
import { Cell, Difficulty } from "types/general_enums";
import { isOccupied, emptyArray } from "./HelperFunctions";

interface Args {
  board: Board;
  currentBoard: GeneralBoardIndex;
  currentDifficulty: Difficulty;
}

interface AiProperties extends Args {
  amountOccupied: NumberBoard;
  avoidBox: NumberBoard;
}

function areTwoValue(value: Cell) {
  if (value === Cell.X) return 90;
  return 40;
}

// Checks which are the more factible actions to play
function different(availableMoves: AiAction[]) {
  const min = availableMoves[0].value;
  return availableMoves.findIndex((action) => action.value > min);
}

function deleteElements(availableMoves: AiAction[]) {
  availableMoves.sort((a, b) => b.value - a.value);
  const erasePos = different(availableMoves);
  if (erasePos !== -1) {
    availableMoves.splice(erasePos, availableMoves.length - erasePos);
  }
  return availableMoves;
}

function getAvailableMoves(props: AiProperties) {
  const availableMoves: AiAction[] = [];
  props.board.forEach((element, index) => {
    if (isOccupied(element)) {
      return;
    }
    const pos = index as GeneralBoardIndex;
    availableMoves.push(
      new AiAction({ pos, boardCopy: props.board, ...props })
    );
  });
  return availableMoves;
}

class Ai {
  board: Board;

  currentBoard: GeneralBoardIndex;

  currentDifficulty: Difficulty;

  amountOccupied = emptyArray();

  avoidBox = emptyArray();

  constructor({ board, currentBoard, currentDifficulty }: Args) {
    this.board = board;
    this.currentDifficulty = currentDifficulty;
    this.currentBoard = currentBoard;
  }

  setVariables = ({ board, currentBoard, currentDifficulty }: Args) => {
    this.board = board;
    this.currentDifficulty = currentDifficulty;
    this.currentBoard = currentBoard;
  };

  getProperties = (): AiProperties => ({
    amountOccupied: this.amountOccupied,
    avoidBox: this.avoidBox,
    board: this.board,
    currentBoard: this.currentBoard,
    currentDifficulty: this.currentDifficulty,
  });

  areTwo = (
    pos1: GeneralBoardIndex,
    pos2: GeneralBoardIndex,
    pos3: GeneralBoardIndex
  ) => {
    const value1 = this.board[pos1];
    const value2 = this.board[pos2];
    const value3 = this.board[pos3];

    const firstIsOccupied = isOccupied(value1);
    const secondIsOccupied = isOccupied(value2);

    if (
      (value1 === value2 && firstIsOccupied && !isOccupied(value3)) ||
      (value1 === value3 && firstIsOccupied && !secondIsOccupied)
    ) {
      return areTwoValue(value1);
    }
    if (value2 === value3 && secondIsOccupied && !firstIsOccupied) {
      return areTwoValue(value2);
    }
    return 0;
  };

  areTwoInTheColumn = () =>
    this.areTwo(0, 3, 6) + this.areTwo(1, 4, 7) + this.areTwo(2, 5, 8);

  areTwoInTheRow = () =>
    this.areTwo(0, 1, 2) + this.areTwo(3, 4, 5) + this.areTwo(6, 7, 8);

  areTwoInTheDiagonal = () => this.areTwo(0, 4, 8) + this.areTwo(2, 4, 6);

  areTwoInTheBoard = () =>
    this.areTwoInTheColumn() +
    this.areTwoInTheDiagonal() +
    this.areTwoInTheRow();

  playerMadeAMove = (boardId: GeneralBoardIndex, board: Board) => {
    this.amountOccupied[boardId] += 3;
    this.board = board;
    this.avoidBox[boardId] = this.areTwoInTheBoard();
  };

  aiMadeAMove = (board: Board) => {
    this.amountOccupied[this.currentBoard] += 1;
    this.board = board;
    this.avoidBox[this.currentBoard] = this.areTwoInTheBoard();
  };

  cleanVariables = () => {
    this.amountOccupied = emptyArray();
    this.avoidBox = emptyArray();
  };

  makeMove = (args: Args) => {
    this.setVariables(args);
    const properties = this.getProperties();
    let availableMoves = getAvailableMoves(properties);
    if (this.currentDifficulty >= Difficulty.NORMAL) {
      availableMoves = deleteElements(availableMoves);
    }
    const index = Math.floor(Math.random() * availableMoves.length);
    return availableMoves[index].pos;
  };
}

export default Ai;
