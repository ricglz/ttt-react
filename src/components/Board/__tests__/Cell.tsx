import renderer from "react-test-renderer";
import type { BigBoard } from "@/types/general";
// eslint-disable-next-line import/extensions
import { Cell as CellEnum } from "@/types/general_enums";
import { newBoard } from "../../../functions/HelperFunctions";
import Cell from "../Cell";

describe("Cell", () => {
  const boardNum = 0;
  const cellNum = 0;
  let boardGame: BigBoard;

  beforeEach(() => {
    boardGame = newBoard();
  });

  test.each([CellEnum.X, CellEnum.O, CellEnum.NONE])("Value is %s", (value) => {
    boardGame[boardNum][cellNum] = value;
    const component = (
      <Cell
        boardNum={boardNum}
        boardGame={boardGame}
        cellNum={cellNum}
        currentBoard={-1}
        handleClick={() => {}}
      />
    );
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
