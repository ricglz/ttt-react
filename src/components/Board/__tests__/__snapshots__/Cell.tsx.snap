// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell Value is -1 1`] = `
<div
  className="col-4 box"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="button"
  tabIndex={-1}
>
  <p
    className="o"
  >
    O
  </p>
</div>
`;

exports[`Cell Value is 0 1`] = `
<div
  className="col-4 box"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="button"
  tabIndex={-1}
>
  <p
    className=""
  >
    
  </p>
</div>
`;

exports[`Cell Value is 1 1`] = `
<div
  className="col-4 box"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="button"
  tabIndex={-1}
>
  <p
    className="x"
  >
    X
  </p>
</div>
`;
